[ICE]: layout_of: unexpected type `_`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
#![feature(transmutability)]
fn assert_transmutable<T>()
where
    (): std::mem::TransmuteFrom<T>,
{
}
fn main() {
    assert_transmutable()
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.94.0-nightly (fe98ddcfc 2026-01-17)
binary: rustc
commit-hash: fe98ddcfcfb6f185dbf4adeaf439d8a756da0273
commit-date: 2026-01-17
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.8
```

### Error output

```

```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: /rustc-dev/fe98ddcfcfb6f185dbf4adeaf439d8a756da0273/compiler/rustc_ty_utils/src/layout.rs:781:13: layout_of: unexpected type `_`


thread 'rustc' (1554896) panicked at /rustc-dev/fe98ddcfcfb6f185dbf4adeaf439d8a756da0273/compiler/rustc_ty_utils/src/layout.rs:781:13:
Box<dyn Any>
stack backtrace:
   0:     0x7fe477e42f93 - <<std[7637d35b2e73f491]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[e59341b65d367131]::fmt::Display>::fmt
   1:     0x7fe47840e2c8 - core[e59341b65d367131]::fmt::write
   2:     0x7fe477e59986 - <std[7637d35b2e73f491]::sys::stdio::unix::Stderr as std[7637d35b2e73f491]::io::Write>::write_fmt
   3:     0x7fe477e19008 - std[7637d35b2e73f491]::panicking::default_hook::{closure#0}
   4:     0x7fe477e366c3 - std[7637d35b2e73f491]::panicking::default_hook
   5:     0x7fe476e0b7fa - std[7637d35b2e73f491]::panicking::update_hook::<alloc[e4cd1c42012959c4]::boxed::Box<rustc_driver_impl[1a3eae45778b1a3a]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7fe477e369a2 - std[7637d35b2e73f491]::panicking::panic_with_hook
   7:     0x7fe476e48161 - std[7637d35b2e73f491]::panicking::begin_panic::<rustc_errors[61541e7795e0e52c]::ExplicitBug>::{closure#0}
   8:     0x7fe476e38586 - std[7637d35b2e73f491]::sys::backtrace::__rust_end_short_backtrace::<std[7637d35b2e73f491]::panicking::begin_panic<rustc_errors[61541e7795e0e52c]::ExplicitBug>::{closure#0}, !>
   9:     0x7fe476e35a59 - std[7637d35b2e73f491]::panicking::begin_panic::<rustc_errors[61541e7795e0e52c]::ExplicitBug>
  10:     0x7fe476e65e61 - <rustc_errors[61541e7795e0e52c]::diagnostic::BugAbort as rustc_errors[61541e7795e0e52c]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
  11:     0x7fe477415df9 - rustc_middle[f49c2f37af7e8c7e]::util::bug::opt_span_bug_fmt::<rustc_span[d2f47af3535269a]::span_encoding::Span>::{closure#0}
  12:     0x7fe477415f82 - rustc_middle[f49c2f37af7e8c7e]::ty::context::tls::with_opt::<rustc_middle[f49c2f37af7e8c7e]::util::bug::opt_span_bug_fmt<rustc_span[d2f47af3535269a]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  13:     0x7fe4774075ab - rustc_middle[f49c2f37af7e8c7e]::ty::context::tls::with_context_opt::<rustc_middle[f49c2f37af7e8c7e]::ty::context::tls::with_opt<rustc_middle[f49c2f37af7e8c7e]::util::bug::opt_span_bug_fmt<rustc_span[d2f47af3535269a]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  14:     0x7fe474ccb234 - rustc_middle[f49c2f37af7e8c7e]::util::bug::bug_fmt
  15:     0x7fe478d7fc51 - rustc_ty_utils[f823f1d5efade7eb]::layout::layout_of_uncached
  16:     0x7fe478d74589 - rustc_ty_utils[f823f1d5efade7eb]::layout::layout_of
  17:     0x7fe478d744f0 - rustc_query_impl[f993d76296ac3ede]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f993d76296ac3ede]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f49c2f37af7e8c7e]::query::erase::Erased<[u8; 16usize]>>
  18:     0x7fe478d735d7 - rustc_query_system[11b36a7129bd2e9f]::query::plumbing::try_execute_query::<rustc_query_impl[f993d76296ac3ede]::DynamicConfig<rustc_query_system[11b36a7129bd2e9f]::query::caches::DefaultCache<rustc_middle[f49c2f37af7e8c7e]::ty::PseudoCanonicalInput<rustc_middle[f49c2f37af7e8c7e]::ty::Ty>, rustc_middle[f49c2f37af7e8c7e]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[f993d76296ac3ede]::plumbing::QueryCtxt, false>
  19:     0x7fe478d731f4 - rustc_query_impl[f993d76296ac3ede]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  20:     0x7fe477d5f028 - rustc_transmute[a79904f2c7e85e92]::layout::rustc::layout_of
  21:     0x7fe477d5d0a2 - <rustc_transmute[a79904f2c7e85e92]::layout::tree::Tree<rustc_transmute[a79904f2c7e85e92]::layout::rustc::Def, rustc_middle[f49c2f37af7e8c7e]::ty::region::Region, rustc_middle[f49c2f37af7e8c7e]::ty::Ty>>::from_ty
  22:     0x7fe477d56d74 - <rustc_transmute[a79904f2c7e85e92]::rustc::TransmuteTypeEnv>::is_transmutable
  23:     0x7fe477c8a775 - <rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate as rustc_next_trait_solver[170a8c4b65e982f7]::delegate::SolverDelegate>::is_transmutable
  24:     0x7fe4771755b0 - <rustc_type_ir[a44b4988878e9ab6]::predicate::TraitPredicate<rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt> as rustc_next_trait_solver[170a8c4b65e982f7]::solve::assembly::GoalKind<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate, rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>::consider_builtin_transmute_candidate
  25:     0x7fe4798ba564 - <rustc_next_trait_solver[170a8c4b65e982f7]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate, rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>::assemble_builtin_impl_candidates::<rustc_type_ir[a44b4988878e9ab6]::predicate::TraitPredicate<rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>
  26:     0x7fe4792c757e - <rustc_next_trait_solver[170a8c4b65e982f7]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate, rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>::assemble_and_evaluate_candidates::<rustc_type_ir[a44b4988878e9ab6]::predicate::TraitPredicate<rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>
  27:     0x7fe479652e29 - <rustc_next_trait_solver[170a8c4b65e982f7]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate, rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>::compute_trait_goal
  28:     0x7fe4792dad22 - <rustc_next_trait_solver[170a8c4b65e982f7]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate> as rustc_type_ir[a44b4988878e9ab6]::search_graph::Delegate>::compute_goal::{closure#0}
  29:     0x7fe4792d9d8d - <rustc_type_ir[a44b4988878e9ab6]::search_graph::SearchGraph<rustc_next_trait_solver[170a8c4b65e982f7]::solve::search_graph::SearchGraphDelegate<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate>, rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>::evaluate_goal::{closure#0}::{closure#2}
  30:     0x7fe47901d8c5 - <rustc_next_trait_solver[170a8c4b65e982f7]::solve::eval_ctxt::EvalCtxt<rustc_trait_selection[ea11fdd225b236a7]::solve::delegate::SolverDelegate, rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>>::evaluate_goal_raw
  31:     0x7fe47903dd11 - <rustc_trait_selection[ea11fdd225b236a7]::solve::fulfill::FulfillmentCtxt<rustc_trait_selection[ea11fdd225b236a7]::traits::FulfillmentError> as rustc_infer[66736765ef873b69]::traits::engine::TraitEngine<rustc_trait_selection[ea11fdd225b236a7]::traits::FulfillmentError>>::try_evaluate_obligations
  32:     0x7fe47875a211 - <rustc_hir_typeck[cda3b2c1a48d49f5]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  33:     0x7fe47877886f - <rustc_hir_typeck[cda3b2c1a48d49f5]::fn_ctxt::FnCtxt>::check_expr_call
  34:     0x7fe4787596f0 - <rustc_hir_typeck[cda3b2c1a48d49f5]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  35:     0x7fe478753afb - <rustc_hir_typeck[cda3b2c1a48d49f5]::fn_ctxt::FnCtxt>::check_expr_block
  36:     0x7fe4787596cb - <rustc_hir_typeck[cda3b2c1a48d49f5]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  37:     0x7fe478739e33 - rustc_hir_typeck[cda3b2c1a48d49f5]::check::check_fn
  38:     0x7fe47917ccf3 - rustc_hir_typeck[cda3b2c1a48d49f5]::typeck_with_inspect::{closure#0}
  39:     0x7fe47917b674 - rustc_query_impl[f993d76296ac3ede]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f993d76296ac3ede]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f49c2f37af7e8c7e]::query::erase::Erased<[u8; 8usize]>>
  40:     0x7fe47856ecaa - rustc_query_system[11b36a7129bd2e9f]::query::plumbing::try_execute_query::<rustc_query_impl[f993d76296ac3ede]::DynamicConfig<rustc_data_structures[ab77da159d78a1fb]::vec_cache::VecCache<rustc_span[d2f47af3535269a]::def_id::LocalDefId, rustc_middle[f49c2f37af7e8c7e]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[11b36a7129bd2e9f]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[f993d76296ac3ede]::plumbing::QueryCtxt, false>
  41:     0x7fe47856e589 - rustc_query_impl[f993d76296ac3ede]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  42:     0x7fe478c14d35 - <rustc_middle[f49c2f37af7e8c7e]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_hir_analysis[5017dacce67ac1a4]::check_crate::{closure#2}>::{closure#0}
  43:     0x7fe478c14744 - rustc_hir_analysis[5017dacce67ac1a4]::check_crate
  44:     0x7fe478562e5b - rustc_interface[70956b6af7f02b7a]::passes::analysis
  45:     0x7fe478562ab5 - rustc_query_impl[f993d76296ac3ede]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[f993d76296ac3ede]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[f49c2f37af7e8c7e]::query::erase::Erased<[u8; 0usize]>>
  46:     0x7fe4795f9e23 - rustc_query_system[11b36a7129bd2e9f]::query::plumbing::try_execute_query::<rustc_query_impl[f993d76296ac3ede]::DynamicConfig<rustc_query_system[11b36a7129bd2e9f]::query::caches::SingleCache<rustc_middle[f49c2f37af7e8c7e]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[f993d76296ac3ede]::plumbing::QueryCtxt, false>
  47:     0x7fe4795f9c0e - rustc_query_impl[f993d76296ac3ede]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  48:     0x7fe479617dd1 - <rustc_interface[70956b6af7f02b7a]::passes::create_and_enter_global_ctxt<core[e59341b65d367131]::option::Option<rustc_interface[70956b6af7f02b7a]::queries::Linker>, rustc_driver_impl[1a3eae45778b1a3a]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[e59341b65d367131]::ops::function::FnOnce<(&rustc_session[ef90395b1d579b60]::session::Session, rustc_middle[f49c2f37af7e8c7e]::ty::context::CurrentGcx, alloc[e4cd1c42012959c4]::sync::Arc<rustc_data_structures[ab77da159d78a1fb]::jobserver::Proxy>, &std[7637d35b2e73f491]::sync::once_lock::OnceLock<rustc_middle[f49c2f37af7e8c7e]::ty::context::GlobalCtxt>, &rustc_data_structures[ab77da159d78a1fb]::sync::worker_local::WorkerLocal<rustc_middle[f49c2f37af7e8c7e]::arena::Arena>, &rustc_data_structures[ab77da159d78a1fb]::sync::worker_local::WorkerLocal<rustc_hir[78cba43770fa9cc1]::Arena>, rustc_driver_impl[1a3eae45778b1a3a]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  49:     0x7fe479598cf4 - rustc_interface[70956b6af7f02b7a]::interface::run_compiler::<(), rustc_driver_impl[1a3eae45778b1a3a]::run_compiler::{closure#0}>::{closure#1}
  50:     0x7fe479535dba - std[7637d35b2e73f491]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[70956b6af7f02b7a]::util::run_in_thread_with_globals<rustc_interface[70956b6af7f02b7a]::util::run_in_thread_pool_with_globals<rustc_interface[70956b6af7f02b7a]::interface::run_compiler<(), rustc_driver_impl[1a3eae45778b1a3a]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  51:     0x7fe479535b7e - <std[7637d35b2e73f491]::thread::lifecycle::spawn_unchecked<rustc_interface[70956b6af7f02b7a]::util::run_in_thread_with_globals<rustc_interface[70956b6af7f02b7a]::util::run_in_thread_pool_with_globals<rustc_interface[70956b6af7f02b7a]::interface::run_compiler<(), rustc_driver_impl[1a3eae45778b1a3a]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[e59341b65d367131]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  52:     0x7fe479537944 - <std[7637d35b2e73f491]::sys::thread::unix::Thread>::new::thread_start
  53:     0x7fe472e9698b - <unknown>
  54:     0x7fe472f1a9cc - <unknown>
  55:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/matthias/vcs/github/CRED/rustc-ice-2026-01-18T16_00_04-1554878.txt` to your bug report

note: rustc 1.94.0-nightly (fe98ddcfc 2026-01-17) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z next-solver=globally

query stack during panic:
panicked at /rustc-dev/fe98ddcfcfb6f185dbf4adeaf439d8a756da0273/compiler/rustc_type_ir/src/ty_kind.rs:680:17:

thread panicked while processing panic. aborting.
```

</p>
</details>

old solver:
```
error[E0283]: type annotations needed
 --> code.rs:8:5
  |
8 |     assert_transmutable()
  |     ^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `T` declared on the function `assert_transmutable`
  |
  = note: cannot satisfy `(): TransmuteFrom<_, Assume { alignment: false, lifetimes: false, safety: false, validity: false }>`
note: required by a bound in `assert_transmutable`
 --> code.rs:4:9
  |
2 | fn assert_transmutable<T>()
  |    ------------------- required by a bound in this function
3 | where
4 |     (): std::mem::TransmuteFrom<T>,
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_transmutable`
help: consider specifying the generic argument
  |
8 |     assert_transmutable::<T>()
  |                        +++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0283`.
```

--- a/tests/ui/traits/next-solver/transmute-from-unconstrained.rs
+++ b/tests/ui/traits/next-solver/transmute-from-unconstrained.rs
@@ -1,0 +1,20 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/151314>.
+//!
+//! Calling a function with an unconstrained `TransmuteFrom` obligation used to
+//! trigger a `layout_of: unexpected type` ICE under the next-gen trait solver
+//! instead of reporting that type annotations are needed.
+
+//@ compile-flags: -Znext-solver=globally
+
+#![feature(transmutability)]
+
+fn assert_transmutable<T>()
+where
+    (): std::mem::TransmuteFrom<T>,
+{
+}
+
+fn main() {
+    assert_transmutable()
+    //~^ ERROR type annotations needed
+}

--- a/tests/ui/traits/next-solver/transmute-from-unconstrained.stderr
+++ b/tests/ui/traits/next-solver/transmute-from-unconstrained.stderr
@@ -1,0 +1,23 @@
+error[E0283]: type annotations needed
+  --> $DIR/transmute-from-unconstrained.rs:18:5
+   |
+LL |     assert_transmutable()
+   |     ^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `T` declared on the function `assert_transmutable`
+   |
+   = note: cannot satisfy `(): TransmuteFrom<_, Assume { alignment: false, lifetimes: false, safety: false, validity: false }>`
+note: required by a bound in `assert_transmutable`
+  --> $DIR/transmute-from-unconstrained.rs:13:9
+   |
+LL | fn assert_transmutable<T>()
+   |    ------------------- required by a bound in this function
+LL | where
+LL |     (): std::mem::TransmuteFrom<T>,
+   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_transmutable`
+help: consider specifying a concrete type for the type parameter `T`
+   |
+LL |     assert_transmutable::</* Type */>()
+   |                        ++++++++++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0283`.

